jkube-kit/build/api/src/main/java/org/eclipse/jkube/kit/build/api/helper/DockerFileUtil.java

```diff
@@ -130,6 +146,54 @@ private static Reader getFileReaderFromDir(File file) {
         }
     }
 
+    /**
+     * Helper method for extractArgs(exposed for test)
+     *
+     * @param argLines list of string arrays containing lines with words
+     * @param argsFromBuildConfig Docker build args from Build Configuration
+     * @return map of parsed arguments
+     */
+    protected static Map<String, String> extractArgsFromLines(List<String[]> argLines, Map<String, String> argsFromBuildConfig) {
+        Map<String, String> result = new HashMap<>();
+        for (String[] argLine : argLines) {
+            if (argLine.length > 1) {
+                updateMapWithArgValue(result, argsFromBuildConfig, argLine[1]);
+            }
+        }
+        return result;
+    }
+
+    private static String resolveImageTagFromArgs(String imageTagString, Map<String, String> args) {
+        if (imageTagString.startsWith("$")) { // FROM $IMAGE
+            String resolvedVal = resolveArgValueFromStrContainingArgKey(imageTagString, args);
+            if (resolvedVal != null) {
+                return resolvedVal;
+            }
+        } else { // FROM image:$TAG_ARG
+            String[] imageTagArr = imageTagString.split(":");
+            if (imageTagArr.length > 1) {
+                String tag = resolveArgValueFromStrContainingArgKey(imageTagArr[1], args);
+                if (tag != null) {
+                    return imageTagArr[0] + ":" + tag;
+                }
+            }
+        }
+        return imageTagString;
+    }
+
+    static String resolveArgValueFromStrContainingArgKey(String argString, Map<String, String> args) {
+        Pattern argPattern = Pattern.compile(ARG_PATTERN_REGEX);
```

Comment:
Declare the pattern as a constant, it's thread-safe. This way you're compiling the pattern  for every method call.